File: palindromic-substrings/Solution647.java
// https://leetcode.com/problems/palindromic-substrings/

class Solution647 {

    // Two pointer, expand around center
    // T: O(n^2), S: O(1)
    public int countSubstrings(String s) {
        int res = 0;
        for (int i = 0; i < s.length(); i++) {
            res += countPalindrome(s, i, i);
            res += countPalindrome(s, i, i + 1);
        }
        return res;
    }

    // T: O(n), S: O(1)
    private int countPalindrome(String s, int l, int r) {
        int res = 0;
        while (l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
            res++;
            l--;
            r++;
        }
        return res;
    }

    // T: O(n^2), S: O(n^2)
    public int countSubstringsDp(String s) {
        int res = 0, n = s.length();
        boolean[][] dp = new boolean[n][n];

        for (int i = n - 1; i >= 0; i--) {
            for (int j = i; j < n; j++) {
                if (s.charAt(i) == s.charAt(j) &&
                        (j - i <= 2 || dp[i + 1][j - 1])) {

                    dp[i][j] = true;
                    res++;
                }
            }
        }

        return res;
    }

    // T: O(n^3), S: O(1)
    public int countSubstringsBrute(String s) {
        int res = 0;

        for (int i = 0; i < s.length(); i++) {
            for (int j = i; j < s.length(); j++) {
                int l = i, r = j;
                while (l < r && s.charAt(l) == s.charAt(r)) {
                    l++;
                    r--;
                }
                res += (l >= r) ? 1 : 0;
            }
        }

        return res;
    }

    public static void main(String[] args) {
        var sol = new Solution647();
        var s1 = "abc";
        var s2 = "aaa";
        System.out.println(sol.countSubstrings(s1)); // 3
        System.out.println(sol.countSubstringsBrute(s1)); // 3
        System.out.println(sol.countSubstringsDp(s1)); // 3
        System.out.println(sol.countSubstrings(s2)); // 6
        System.out.println(sol.countSubstringsDp(s2)); // 6
        System.out.println(sol.countSubstringsBrute(s2)); // 6
    }

}

File: palindromic-substrings/solution647.py
# https://leetcode.com/problems/palindromic-substrings/

class Solution647:

    # Two pointer, expand around center
    # T: O(n^2), S: O(1)
    def countSubstrings(self, s: str) -> int:
        res = 0
        for i in range(len(s)):
            res += self.count_palindrome(s, i, i)
            res += self.count_palindrome(s, i, i + 1)
        return res

    # T: O(n), S: O(1)
    def count_palindrome(self, s: str, l: int, r: int) -> int:
        res = 0
        while l >= 0 and r < len(s) and s[l] == s[r]:
            res += 1
            l -= 1
            r += 1
        return res

    # T: O(n^2), S: O(n^2)
    def countSubstringsDp(self, s: str) -> int:
        n = len(s)
        res = 0
        dp = [[False] * n for _ in range(n)]

        for i in range(n - 1, -1, -1):
            for j in range(i, n):
                if s[i] == s[j] and (j - i <= 2 or dp[i + 1][j - 1]):
                    dp[i][j] = True
                    res += 1

        return res

    # T: O(n^3), S: O(1)
    def countSubstringsBrute(self, s: str) -> int:
        res = 0

        for i in range(len(s)):
            for j in range(i, len(s)):
                l, r = i, j
                while l < r and s[l] == s[r]:
                    l += 1
                    r -= 1
                if l >= r:
                    res += 1

        return res


def main() -> None:
    sol = Solution647()
    s1 = "abc"
    s2 = "aaa"
    print(sol.countSubstrings(s1))      # 3
    print(sol.countSubstringsBrute(s1))  # 3
    print(sol.countSubstringsDp(s1))    # 3
    print(sol.countSubstrings(s2))      # 6
    print(sol.countSubstringsDp(s2))    # 6
    print(sol.countSubstringsBrute(s2))  # 6


if __name__ == "__main__":
    main()
